### PR TITLE
Fall Cleaning

### LINKED ISSUE #1282: proposed deprecations
- [ ] subinplace
- [x] zero_lower_unbounded
- [ ] markdown (low priority)
- [x] composite objective

### LINKED ISSUE #1288: ConstraintSets which inherit from dicts
so that subconstraints are labeled

### LINKED ISSUE #1358: make signomial 'tautological constraint' error a warning
It seems like there is a workaround for posynomials, but is there one for signomials? Let me know @bqpd.

### LINKED ISSUE #1363: Environment path length issue
I figured out the reason for the issue. It's due to the following lines in gpkit/_mosek/cli_expopt.py, on line 38.
```python
    if "mosek_bin_dir" in settings:
        os.environ['PATH'] = ":".join([os.environ['PATH'],
                                       settings["mosek_bin_dir"]])
```
This for some reason makes the length of PATH so long that it breaks some of the sweeps I am running for robustSP stuff. I changed it to the following in my version of gpkit.
```python
    if "mosek_bin_dir" in settings:
        os.environ['PATH'] = ":".join([os.environ['PATH']])
```
to avoid the growing PATH length. Do you have insight into this @bqpd ?

### PR DESCRIPTION
**Pull request by bqpd**
Closes #1282
Closes #1363 
Closes #1358
Closes #1288

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,6 +1,9 @@
 # iPython checkpoint folders
 .ipynb_checkpoints/
 
+# Pickled solutions
+*.p
```
Comment (whoburg) [inline .gitignore]:
```suggestion
*.pkl
```

.pkl is extension suggested by python 2 docs: 
https://docs.python.org/2/library/pickle.html#example

File: docs/source/examples/performance_modeling.py [hunk 2]
```diff
@@ -217,6 +238,9 @@ def setup(self):
 sol_loaded = pickle.load(open("solution.p"))
 
 vars_of_interest = set(AC.varkeys)
+# note that there's two ways to access submodels
+assert (MISSION["flight segment"]["aircraft performance"]
+        is MISSION.fs.aircraftp)
```
Comment (whoburg) [inline docs/source/examples/performance_modeling.py]: Nice

File: gpkit/_mosek/cli_expopt.py [hunk 3]
```diff
@@ -36,16 +36,16 @@ def imize_fn(path=None, clearfiles=True):
         path = tempfile.mkdtemp()
     filename = path + os.sep + "gpkit_mosek"
     if "mosek_bin_dir" in settings:
-        os.environ['PATH'] = ":".join([os.environ['PATH'],
-                                       settings["mosek_bin_dir"]])
+        if settings["mosek_bin_dir"] not in os.environ["PATH"]:
+            os.environ["PATH"] = (os.environ["PATH"]
+                                  + ":" + settings["mosek_bin_dir"])
```
Comment (whoburg) [inline gpkit/_mosek/cli_expopt.py]: good fix... 

`:.join(` seemed cleaner to me than +, just curious, was it intentional to switch?

File: gpkit/_pint/__init__.py [hunk 4]
```diff
@@ -25,13 +26,17 @@ def qty(unit):
...
-    "Return monomials instead of Quantitites"
+    "Return Monomials instead of Quantitites"
 
     def __call__(self, arg):
+        "Returns a unit Monomial, caching the result for future retrievals"
         from .. import Monomial
-        return Monomial(qty(arg))
+        if arg not in MON_CACHE:
+            MON_CACHE[arg] = Monomial(qty(arg))
+        return MON_CACHE[arg]
```
Comment (whoburg) [inline gpkit/_pint/__init__.py]: Nice. Would putting the import inside the conditional speed this up further, or is that silly?

File: gpkit/constraints/set.py [hunk 5]
```diff
@@ -36,8 +36,13 @@ class ConstraintSet(list):
     "Recursive container for ConstraintSets and Inequalities"
     varkeys = None
     unique_varkeys = frozenset()
+    # idxlookup holds the names of the top-level constraintsets
```
Comment (whoburg) [inline gpkit/constraints/set.py]: 👏 inline comment

File: gpkit/constraints/set.py [hunk 6]
```diff
@@ -92,6 +97,8 @@ def __init__(self, constraints, substitutions=None):  # pylint: disable=too-many
         add_meq_bounds(self.bounded, self.meq_bounded)
 
     def __getitem__(self, key):
+        if self.idxlookup and key in self.idxlookup:
+            key = self.idxlookup[key]
```
Comment (whoburg) [inline gpkit/constraints/set.py]: Really feels like this thing should be a collections.OrderedDict, but I haven't thought deeply about it

File: gpkit/constraints/set.py [hunk 7]
```diff
@@ -220,34 +229,6 @@ def flat(self, constraintsets=True):
...
-        if subkeys.intersection(self.substitutions):
-            for key, value in subs.items():
-                if key in self.substitutions:
-                    valkey, _ = self.substitutions.parse_and_index(value)
-                    self.substitutions[valkey] = self.substitutions[key]
-                    del self.substitutions[key]
-        self.unique_varkeys = frozenset(subs[vk] if vk in subs else vk
-                                        for vk in self.unique_varkeys)
-        self.reset_varkeys()
-
```
Comment (whoburg) [inline gpkit/constraints/set.py]: Hooray!!

File: gpkit/tests/t_model.py [hunk 8]
```diff
@@ -352,16 +352,54 @@ def test_sp_substitutions(self):
...
+
+    def test_tautological(self):
+        x = Variable('x')
+        y = Variable('y')
+        z = Variable('z')
+
+        import sys
+        from cStringIO import StringIO
+        old_stdout = sys.stdout
+        sys.stdout = stringout = StringIO()
```
Comment (whoburg) [inline gpkit/tests/t_model.py]: Duplicated in another method above, belongs in 
```suggestion
def setUp(self):
``` 
?
Also, inline comment?

File: gpkit/tests/t_model.py [hunk 9]
```diff
@@ -352,16 +352,54 @@ def test_sp_substitutions(self):
...
+        m2.substitutions.update({'y': 1})
+        self.assertAlmostEqual(m1.solve(self.solver, verbosity=0)["cost"],
+                               m2.solve(self.solver, verbosity=0)["cost"])
+
+        sys.stdout = old_stdout
+        self.assertEqual(stringout.getvalue(), (
+            "Warning: SignomialConstraint x + y >= z became the tautological"
+            " constraint 0 <= 1 + x after substitution.\n"
+            "Warning: SignomialConstraint 1 + x >= 0 became the tautological"
+            " constraint 0 <= 1 + x after substitution.\n"))
```
Comment (whoburg) [inline gpkit/tests/t_model.py]: Nice

**Comment by bqpd** [comment]
test this please

**Comment by bqpd** [comment]
test models please

File: gpkit/_mosek/cli_expopt.py (hunk 3 above)
Comment (bqpd) [reply gpkit/_mosek/cli_expopt.py]: Yeah, I think it's less confusing to use "+" when it's always a two-element list.

File: gpkit/_pint/__init__.py (hunk 4 above)
Comment (bqpd) [reply gpkit/_pint/__init__.py]: naw it's cached by the interpreter and basically free

**Comment by bqpd** [comment]
test models please

**Pull request merged**